armi/reactor/geometry.py

```diff
@@ -113,6 +123,308 @@ def __str__(self):
+        (DomainType.FULL_CORE, BoundaryType.NO_SYMMETRY, True),
+        (DomainType.THIRD_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.QUARTER_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.QUARTER_CORE, BoundaryType.REFLECTIVE, False),
+        (DomainType.QUARTER_CORE, BoundaryType.PERIODIC, True),
+        (DomainType.QUARTER_CORE, BoundaryType.REFLECTIVE, True),
+        (DomainType.EIGHTH_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.EIGHTH_CORE, BoundaryType.REFLECTIVE, False),
+        (DomainType.EIGHTH_CORE, BoundaryType.PERIODIC, True),
+        (DomainType.EIGHTH_CORE, BoundaryType.REFLECTIVE, True),
+        (DomainType.SIXTEENTH_CORE, BoundaryType.PERIODIC, False),
+        (DomainType.SIXTEENTH_CORE, BoundaryType.REFLECTIVE, False),
+    }
+
+    @classmethod
+    def validSymmetryStrings(cls):
+        """Convert VALID_SYMMETRY tuples to strings."""
+        return [str(cls.fromAny(x)) for x in cls.VALID_SYMMETRY]
+
+    @staticmethod
+    def _checkIfThroughCenter(centerString: str):
+        return THROUGH_CENTER_ASSEMBLY in centerString
+
+    def __init__(
+        self,
+        domainType: "DomainType" = DomainType.THIRD_CORE,
+        boundaryType: "BoundaryType" = BoundaryType.PERIODIC,
+        throughCenterAssembly: Optional[bool] = False,
+    ):
+
+        if (domainType, boundaryType, throughCenterAssembly) in self.VALID_SYMMETRY:
+            self.domain = domainType
+            self.boundary = boundaryType
+            self.isThroughCenterAssembly = throughCenterAssembly
+        else:
+            symmetryString = [str(domainType)]
+            if not boundaryType == BoundaryType.NO_SYMMETRY:
+                symmetryString.append(str(boundaryType))
+            if throughCenterAssembly:
+                symmetryString.append(THROUGH_CENTER_ASSEMBLY)
+            errorMsg = (
+                "{} is not a valid symmetry option. Valid symmetry options are:".format(
+                    " ".join(symmetryString)
+                )
+            )
+            errorMsg += ", ".join([f"{sym}" for sym in self.validSymmetryStrings()])
+            raise ValueError(errorMsg)
+
+    @classmethod
+    def fromStr(cls, symmetryString: str) -> "SymmetryType":
+        isThroughCenter = cls._checkIfThroughCenter(symmetryString)
+        coreString = symmetryString.replace(THROUGH_CENTER_ASSEMBLY, "")
+        pieces = coreString.split()
+        domain = DomainType.fromStr(pieces[0])
+        if len(pieces) <= 1 or pieces[-1] == "core":
+            # set the BoundaryType to a default for the DomainType
+            if domain == DomainType.FULL_CORE:
+                boundary = BoundaryType.NO_SYMMETRY
+            elif domain == DomainType.THIRD_CORE:
+                boundary = BoundaryType.PERIODIC
+            else:
+                boundary = BoundaryType.REFLECTIVE
+        else:
+            boundary = BoundaryType.fromStr(pieces[-1])
```

Comment:
My only beef with this is that (I think) it would allow parsing the string "quarter [entire works of Shakespeare] periodic" with no complaint. Will this probably ever happen in practice? No, probably not. I know it's kind of pedantic, but it changes the semantics of this function from "convert a set of valid strings into a valid pair enumeration values, or `raise`" to "Try to extract a valid pair of enumeration values from a string, sometimes `raise`ing if the individual attempts at parsing parts of that string into their respective enum values doesn't work." In many cases, both of these behaviors will accomplish the same thing, but the first one is simpler, and more importantly what someone would reasonably _expect_ the semantics to be.

Reply:
I modified it again to explicitly look for `"core"` and remove it if it's there. Then we can just interpret the rest as a string with exactly one or two words, and if it's anything but one or two valid words it will throw an error.